metadata-ingestion/src/datahub/ingestion/source/state/redundant_run_skip_handler.py

```diff
@@ -104,46 +99,178 @@ def create_checkpoint(self) -> Optional[Checkpoint[BaseUsageCheckpointState]]:
+                # scenario of scheduled ingestions with default start, end times
+                suggested_start_time_millis = last_run_end_time
+                logger.info(
+                    f"{self.job_id} : Reducing time window. Changing start time to  {get_datetime_from_ts_millis_in_utc(last_run_end_time)}"
+                )
+        else:
+            # cur_start_time_millis < last_run_start_time
+            # This is most likely a manual backdated run which we should always run.
+            # Do we really need below optimisation/reducing time window for manual runs ?
+            if (
+                last_run_start_time < cur_end_time_millis <= last_run_end_time
+                and allow_reduce
+            ):
+                suggested_end_time_millis = last_run_start_time
+                logger.info(
+                    f"{self.job_id} : Reducing time window. Changing end time to  {get_datetime_from_ts_millis_in_utc(last_run_end_time)}"
+                )
+
+        logger.info(
+            f"{self.job_id} : Suggested start, end times: "
+            f"({get_datetime_from_ts_millis_in_utc(suggested_start_time_millis)}, {get_datetime_from_ts_millis_in_utc(suggested_end_time_millis)})"
+        )
+        return (
+            get_datetime_from_ts_millis_in_utc(suggested_start_time_millis),
+            get_datetime_from_ts_millis_in_utc(suggested_end_time_millis),
+        )
+
+    def get_last_run_time_window(
+        self, last_checkpoint: Checkpoint[BaseTimeWindowCheckpointState]
+    ) -> Tuple[int, int]:
+        # Determine from the last check point state
+        last_run_start_time: int = last_checkpoint.state.begin_timestamp_millis
+        last_run_end_time: int = last_checkpoint.state.end_timestamp_millis
+
+        # For time bucket based aggregations - e.g. usage
+        if last_checkpoint.state.bucket_duration is not None:
+            last_run_end_time = datetime_to_ts_millis(
+                get_time_bucket(
+                    get_datetime_from_ts_millis_in_utc(last_run_end_time),
+                    last_checkpoint.state.bucket_duration,
+                )
             )
-            logger.warning(warn_msg)
-            return True
-        return False
+
+        logger.debug(
+            f"{self.job_id} : Last run start, end times:"
+            f"({get_datetime_from_ts_millis_in_utc(last_run_start_time)}, {get_datetime_from_ts_millis_in_utc(last_run_end_time)})"
+        )
+
+        return last_run_start_time, last_run_end_time
+
+
+class RedundantLineageRunSkipHandler(RedundantRunSkipHandler):
+    def get_job_name_suffix(self):
+        return "_lineage"
+
+
+class RedundantUsageRunSkipHandler(RedundantRunSkipHandler):
+    def get_job_name_suffix(self):
+        return "_usage"
+
+    def update_state(
+        self,
+        start_time: datetime,
+        end_time: datetime,
+        bucket_duration: Optional[BucketDuration] = None,
```

Comment:
Perhaps we can change the type signature here so mypy can catch errors and remove assertion; same below

Reply:
I'm not sure how to get around this. This is an overridden method and hence signature should be same as original method, right ?

Reply:
Ah nvm, you're right. Function parameters are contravariant, i.e. the child can relax a type constraint of its parameter, not restrict it. Can ignore me here

Reply:
done.